File: native/swift/swift-export-standalone-integration-tests/simple/testData/execution/sealedTypes/nesting/nesting.kt
// KIND: STANDALONE
// MODULE: SealedNesting
// FILE: nested_root.kt

// The root and its inheritors sit at the bottom of a `class -> interface -> object` chain, which
// Swift Export flattens into one declaration; the cases must stay reachable through the nested path.

class Outer {
    interface Middle {
        object Inner {
            sealed interface Root

            class Leaf : Root {
                override fun toString(): String = "Leaf"
            }

            object LeafObject : Root {
                override fun toString(): String = "LeafObject"
            }

            /** Not exported: forces the `unknown` case. */
            internal class Hidden : Root {
                override fun toString(): String = "Hidden"
            }
        }
    }
}

fun createNestedLeaf(): Outer.Middle.Inner.Root = Outer.Middle.Inner.Leaf()

fun createNestedLeafObject(): Outer.Middle.Inner.Root = Outer.Middle.Inner.LeafObject

fun createNestedHidden(): Outer.Middle.Inner.Root = Outer.Middle.Inner.Hidden()

// FILE: split_inheritors.kt

// A root nested two levels down whose inheritors sit at *different* nesting sites: top level, a
// deeper container elsewhere, and a non-exported top-level class.

object Holder {
    interface Nested {
        sealed interface SplitRoot
    }
}

class SplitTopLevel : Holder.Nested.SplitRoot {
    override fun toString(): String = "SplitTopLevel"
}

object OtherHolder {
    class Deeper {
        class SplitDeeper : Holder.Nested.SplitRoot {
            override fun toString(): String = "SplitDeeper"
        }
    }
}

/** Not exported. */
internal class SplitHidden : Holder.Nested.SplitRoot {
    override fun toString(): String = "SplitHidden"
}

fun createSplitTopLevel(): Holder.Nested.SplitRoot = SplitTopLevel()

fun createSplitDeeper(): Holder.Nested.SplitRoot = OtherHolder.Deeper.SplitDeeper()

fun createSplitHidden(): Holder.Nested.SplitRoot = SplitHidden()

File: native/swift/swift-export-standalone-integration-tests/simple/testData/execution/sealedTypes/nesting/nesting.swift
import KotlinRuntime
import SealedNesting
import Testing

// Nested scopes are flattened into one generated declaration, but the nested path a client writes
// stays available through typealiases.
private typealias NestedRoot = Outer.Middle.Inner.Root
private typealias SplitRoot = Holder.Nested.SplitRoot

private func describeNested(_ value: any NestedRoot) -> String {
    switch value.sealedType() {
        case let .leaf(type): "leaf: \(type.value)"
        case let .leafObject(type): "leafObject: \(type.value)"
        case let .unknown(type): "unknown: \(type.value)"
    }
}

@Test
func testInheritorsNestedInTheRootsOwnContainer() throws {
    try #require(describeNested(createNestedLeaf()) == "leaf: Leaf")
    try #require(describeNested(createNestedLeafObject()) == "leafObject: LeafObject")
    // The leaves are also constructible through that same nested path.
    try #require(describeNested(Outer.Middle.Inner.Leaf()) == "leaf: Leaf")
    try #require(describeNested(Outer.Middle.Inner.LeafObject.shared) == "leafObject: LeafObject")
}

@Test
func testNonExportedNestedInheritorReportsUnknown() throws {
    try #require(describeNested(createNestedHidden()) == "unknown: Hidden")
}

private func describeSplit(_ value: any SplitRoot) -> String {
    switch value.sealedType() {
        case let .splitTopLevel(type): "topLevel: \(type.value)"
        case let .splitDeeper(type): "deeper: \(type.value)"
        case let .unknown(type): "unknown: \(type.value)"
    }
}

@Test
func testTopLevelInheritorOfANestedRoot() throws {
    try #require(describeSplit(createSplitTopLevel()) == "topLevel: SplitTopLevel")
}

@Test
func testInheritorNestedOutsideTheRootsContainer() throws {
    try #require(describeSplit(createSplitDeeper()) == "deeper: SplitDeeper")
    try #require(describeSplit(OtherHolder.Deeper.SplitDeeper()) == "deeper: SplitDeeper")
}

@Test
func testNonExportedTopLevelInheritorReportsUnknown() throws {
    try #require(describeSplit(createSplitHidden()) == "unknown: SplitHidden")
}
